In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
pd.__version__

'2.0.3'

## Answer1 = '2.0.3'

In [3]:
df = pd.read_csv('data.csv')


## Question2

How many columns are in the dataset?



In [4]:
len(df.columns)

10

## Answer 2 = 10

## Question 3
Which columns in the dataset have missing values?



In [5]:
df.isnull().sum()

longitude               0
latitude                0
housing_median_age      0
total_rooms             0
total_bedrooms        207
population              0
households              0
median_income           0
median_house_value      0
ocean_proximity         0
dtype: int64

## Answer 3 = total_bedrooms

## Question 4
How many unique values does the ocean_proximity column have?

In [6]:
df.ocean_proximity.nunique()


5

## Answer 4 = 5

## Question 5
What's the average value of the median_house_value for the houses located near the bay?



In [7]:
df

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0        -122.23     37.88                41.0        880.0           129.0   
1        -122.22     37.86                21.0       7099.0          1106.0   
2        -122.24     37.85                52.0       1467.0           190.0   
3        -122.25     37.85                52.0       1274.0           235.0   
4        -122.25     37.85                52.0       1627.0           280.0   
...          ...       ...                 ...          ...             ...   
20635    -121.09     39.48                25.0       1665.0           374.0   
20636    -121.21     39.49                18.0        697.0           150.0   
20637    -121.22     39.43                17.0       2254.0           485.0   
20638    -121.32     39.43                18.0       1860.0           409.0   
20639    -121.24     39.37                16.0       2785.0           616.0   

       population  households  median_income  median_house_value  \
0           322.0       126.0         8.3252            452600.0   
1          2401.0      1138.0         8.3014            358500.0   
2           496.0       177.0         7.2574            352100.0   
3           558.0       219.0         5.6431            341300.0   
4           565.0       259.0         3.8462            342200.0   
...           ...         ...            ...                 ...   
20635       845.0       330.0         1.5603             78100.0   
20636       356.0       114.0         2.5568             77100.0   
20637      1007.0       433.0         1.7000             92300.0   
20638       741.0       349.0         1.8672             84700.0   
20639      1387.0       530.0         2.3886             89400.0   

      ocean_proximity  
0            NEAR BAY  
1            NEAR BAY  
2            NEAR BAY  
3            NEAR BAY  
4            NEAR BAY  
...               ...  
20635          INLAND  
20636          INLAND  
20637          INLAND  
20638          INLAND  
20639          INLAND  

[20640 rows x 10 columns]

In [8]:
mean_value = df.loc[df['ocean_proximity'] == 'NEAR BAY', 'median_house_value'].mean()


In [9]:
mean_value

259212.31179039303

## Answer 5 = 259212

## Question 6
Calculate the average of total_bedrooms column in the dataset.
Use the fillna method to fill the missing values in total_bedrooms with the mean value from the previous step.
Now, calculate the average of total_bedrooms again.
Has it changed?


In [10]:
df.total_bedrooms.mean()

537.8705525375618

In [11]:
df_na = df

In [12]:
df_na.isnull().sum()

longitude               0
latitude                0
housing_median_age      0
total_rooms             0
total_bedrooms        207
population              0
households              0
median_income           0
median_house_value      0
ocean_proximity         0
dtype: int64

In [13]:
df_na.total_bedrooms = df_na.total_bedrooms.fillna(df.total_bedrooms.mean())

In [14]:
df_na.total_bedrooms.mean()

537.8705525375617

In [15]:
df_na.total_bedrooms.mean() == df.total_bedrooms.mean()

True

## Answer 6 = Yes

## Question 7

In [16]:
df_na.isnull().sum()

longitude             0
latitude              0
housing_median_age    0
total_rooms           0
total_bedrooms        0
population            0
households            0
median_income         0
median_house_value    0
ocean_proximity       0
dtype: int64

In [29]:
df.ocean_proximity.nunique()

5

In [30]:
df.ocean_proximity.unique()

array(['NEAR BAY', '<1H OCEAN', 'INLAND', 'NEAR OCEAN', 'ISLAND'],
      dtype=object)

In [32]:
selected = df_na[df_na['ocean_proximity']== 'ISLAND']

In [18]:
column_selected =  ['housing_median_age', 'total_rooms', 'total_bedrooms']

In [33]:
X = selected[column_selected].values

In [34]:
XTX = X.T.dot(X)


In [35]:
XTX

array([[9.6820000e+03, 3.5105300e+05, 9.1357000e+04],
       [3.5105300e+05, 1.4399307e+07, 3.7720360e+06],
       [9.1357000e+04, 3.7720360e+06, 9.9835800e+05]])

In [36]:
XTX_inv = np.linalg.inv(XTX)


In [37]:
XTX_inv


array([[ 9.19403586e-04, -3.66412216e-05,  5.43072261e-05],
       [-3.66412216e-05,  8.23303633e-06, -2.77534485e-05],
       [ 5.43072261e-05, -2.77534485e-05,  1.00891325e-04]])

In [38]:
y= [950, 1300, 800, 1000, 1300];


In [39]:
XTX.shape

(3, 3)

In [40]:
XTX_inv.shape

(3, 3)

In [41]:
X.T.shape

(3, 5)

In [42]:
X.shape

(5, 3)

In [49]:

w = XTX_inv.dot(X.T).dot(y)

In [51]:
w[-1]

5.699229455065565

## Answer 7 : 5.699229455065565